kong/plugins/key-auth/handler.lua

```diff
@@ -17,13 +17,20 @@ local KeyAuthHandler = BasePlugin:extend()
 
 KeyAuthHandler.PRIORITY = 1000
 
+local OPTIONS = "OPTIONS"
+
 function KeyAuthHandler:new()
   KeyAuthHandler.super.new(self, "key-auth")
 end
 
 function KeyAuthHandler:access(conf)
   KeyAuthHandler.super.access(self)
 
+  -- do not authenticate preflight OPTIONS requests
+  if ngx.req.get_method() == OPTIONS then
+    return
+  end
+
```

Comment:
In principle, I don't think OPTIONS requests [should be authenticated(https://stackoverflow.com/questions/15734031/why-does-the-preflight-options-request-of-an-authenticated-cors-request-work-in/15734032#15734032). In practice, upstream APIs probably handle OPTIONS differently, so maybe there's a case for making this configurable. I'd be interested in other opinions here, I'm not sure!